subprojects/core/src/main/java/org/gradle/api/internal/artifacts/dsl/dependencies/DependenciesExtensionModule.java

```diff
@@ -239,22 +242,83 @@ public static <D extends Dependency> void call(DependencyCollector self, D depen
         self.add(dependency, ConfigureUtil.configureUsing(configuration));
     }
 
+    @VisibleForTesting
+    public static final String ERROR_MESSAGE_PROVIDER = "Providers of non-Dependency(Constraint) types (java.lang.String) are not supported. Create a Dependency(Constraint) using Dependency(Constraint)Factory first.";
+
     /**
-     * Add a dependency.
+     * Add a dependency or dependency constraint.
      *
-     * @param dependency dependency to add
+     * <p>
+     * Unfortunately, this method cannot be split into two methods, one for {@link Dependency} and one for {@link DependencyConstraint},
+     * because Groovy cannot distinguish between the two.
+     * </p>
+     *
+     * @param dependencyOrDependencyConstraint dependency or dependency constraint to add
      */
-    public static void call(DependencyCollector self, Provider<? extends Dependency> dependency) {
-        self.add(dependency);
+    public static void call(DependencyCollector self, Provider<?> dependencyOrDependencyConstraint) {
```

Comment:
There may be an alternative here where we can inspect the type that the Provider knows (`ProviderInternal.getType`) and when it's unknown, we assume it's one type or just fail. 

The other alternative would be to just call `get` and inspect the type.  That means we break the provider in Groovy DSL.

Reply:
I don't think having an earlier error is worth much. You're already in the DSL at this point, so it should become obvious soon if something's total broken and you can fix it.

Reply:
I think there are a few scenarios where we end up in this method with constraints:
1. Someone is using version catalog with `implementation(constraint(libs.foo))`
2. Someone is creating `Provider<DependencyConstraint>` like `implementation(provider { dependencyConstraintFactory.create(...) })`
3. Someone is mapping and transforming something into a constraint: `implementation(stringProvider.map { constraint(it) })`

I think (1) is the most common. I think you run into this when you just use a version catalog. I don't like that it effectively doubles what each dependency or dependency constraint does.

I think that means:
```
implementation constraint(libs.foo)
implementation libs.bar
```
Both of these are a `Provider` of something and we'll end up getting 2 dependencies and 2 constraint "elements" for the configuration. Each element is a 0-or-1 list and when we resolve it, we'll only get one dependency and one constraint. 

I'd like to try:
1. If the provider knows which type it is, we should short circuit and do the most correct thing. We shouldn't add a dependency and a constraint.
2. If the provider does not know which type it is, we can fallback to adding both.
3. If the provider comes from a method we control (`constraint(...)` or dependency modifier), the provider should know its type.

I think with (1) + (3), we're only going to fall into the workaround when someone is dealing with ad-hoc providers (`project.provider { ... }`) or a mapped provider. Maybe we can improve those in the future so there are fewer cases where Provider.getType is null.

Reply:
Added.